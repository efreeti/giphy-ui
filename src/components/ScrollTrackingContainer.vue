<template>
  <div>
    <slot />

    <div v-if="trackScrollEnd">
      <div class="scroll-end-tracker" role="scroll-end-tracker" v-observe-visibility="handleVisibilityChange"></div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Emit, Vue } from 'vue-property-decorator';

@Component
export default class ScrollTrackingContainer extends Vue {
  @Prop() public trackScrollEnd!: boolean;

  @Emit() scrollEnd() {}

  handleVisibilityChange(isVisible: boolean) {
    if (isVisible) {
      this.scrollEnd();
    }
  }
}
</script>

<style>
.scroll-end-tracker {
  height: 20px;
}
</style>
