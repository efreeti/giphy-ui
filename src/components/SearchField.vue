<template>
  <md-field>
    <label><slot name="label"/></label>

    <md-input :value="value" @input="input" @keyup.enter="search"/>

    <md-button class="md-icon-button md-dense" @click="search">
      <md-icon>search</md-icon>
    </md-button>
  </md-field>
</template>

<script lang="ts">
import { Component, Prop, Emit, Vue } from 'vue-property-decorator';

@Component
export default class SearchField extends Vue {
  @Prop() value!: string;

  @Emit() input() {}

  @Emit() search() {}
}
</script>
