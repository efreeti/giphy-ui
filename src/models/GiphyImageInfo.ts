export interface GiphyImageInfo {
  url: string,
  width: number,
  height: number,
}
