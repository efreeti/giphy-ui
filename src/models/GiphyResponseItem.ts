import { GiphyImageInfo } from "@/models/GiphyImageInfo";

export interface GiphyResponseItem {
  id: string,
  images: {
    downsized_large: GiphyImageInfo,
  },
}
