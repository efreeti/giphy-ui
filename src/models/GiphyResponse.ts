import { GiphyResponseItem } from "@/models/GiphyResponseItem";

export interface GiphyResponse {
  data: Array<GiphyResponseItem>,
  pagination: {
    total_count: number,
  },
}
