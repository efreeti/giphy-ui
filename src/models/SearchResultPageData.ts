import { GiphyResponseItem } from "@/models/GiphyResponseItem";

export class SearchResultPageData {
  constructor(
    public loaded: boolean,
    public images: Array<GiphyResponseItem>,
    public error: any,
  ) {}
}
