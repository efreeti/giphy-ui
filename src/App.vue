<template>
  <md-app md-mode="fixed">
    <md-app-toolbar class="md-layout md-alignment-center-center">
      <div class="md-layout-item md-xlarge-size-30 md-large-size-40 md-medium-size-50 md-small-size-80 md-xsmall-size-100">
        <search-field :value="searchText" @input="updateSearchText" @search="updateSearchResults">
          <template slot="label">Search keyword(s)</template>
        </search-field>
      </div>
    </md-app-toolbar>
    <md-app-content class="md-layout md-alignment-center-center">
    </md-app-content>
  </md-app>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { namespace } from 'vuex-class';

import { SearchResultPageData } from '@/models/SearchResultPageData';

import SearchField from "@/components/SearchField.vue";

const app = namespace('app');

@Component({
  components: {
    'search-field': SearchField,
  },
})
export default class App extends Vue {
  @app.State private searchText!: string;

  @app.State private searchResultPages!: Array<SearchResultPageData>;

  @app.State private searchResultPageSize!: number;

  @app.Mutation private updateSearchText!: (searchText: string) => void;

  @app.Action private updateSearchResults!: () => void;

  @app.Action private showOneMorePage!: () => void;

  @app.Getter private hasMoreSearchResultPages!: () => boolean;

  created() {
    this.updateSearchResults();
  }
}
</script>

<style>
.md-app {
  max-height: 100%;
}
</style>
