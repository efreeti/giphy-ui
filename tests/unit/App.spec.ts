import Vue from 'vue';
import Vuex from 'vuex';
import { mount } from '@vue/test-utils';
import { intersectionObserver } from '@shopify/jest-dom-mocks';

import './config';
import { createImages } from "./utils";

import { SearchResultPageData } from "@/models/SearchResultPageData";
import App from '@/App.vue';
import SearchField from "@/components/SearchField.vue";
import ScrollTrackingContainer from "@/components/ScrollTrackingContainer.vue";

Vue.use(Vuex);

Math.random = () => 0.12345;

const defaultState = {
  searchText: '',
  searchResultPages: [],
  totalSearchResultPages: 0,
};

function createMockedStore(state: any) {
  return {
    modules: {
      app: {
        namespaced: true,
        state,
        getters: {
          hasMoreSearchResultPages: () => state.searchResultPages.length < state.totalSearchResultPages,
        },
        mutations: {
          updateSearchText: jest.fn(),
        },
        actions: {
          updateSearchResults: jest.fn(),
          showOneMorePage: jest.fn(),
        },
      },
    },
  };
}

describe('App', () => {
  beforeEach(() => {
    intersectionObserver.mock();
  });

  afterEach(() => {
    intersectionObserver.restore();
  });

  it('should render correctly', () => {
    const store = new Vuex.Store(createMockedStore({
      searchText: 'text',
      searchResultPages: [
        new SearchResultPageData(true, createImages([1, 2, 3, 4, 5]), null),
        new SearchResultPageData(true, [], new Error('some error')),
        new SearchResultPageData(false, [], null),
      ],
      searchResultPageSize: 10,
      totalSearchResultPages: 4,
    }));

    expect(mount(App, { store })).toMatchSnapshot();
  });

  it('should updateSearchText when editing search field', async () => {
    const store = createMockedStore(defaultState);
    const app = mount(App, { store: new Vuex.Store(store) });

    app.find(SearchField).vm.$emit('input', 'text');

    expect(store.modules.app.mutations.updateSearchText).toHaveBeenCalledWith(defaultState, 'text');
  });

  it('should updateSearchResults when mounted', async () => {
    const store = createMockedStore(defaultState);

    mount(App, { store: new Vuex.Store(store) });

    expect(store.modules.app.actions.updateSearchResults).toHaveBeenCalledWith(
      expect.anything(), undefined, undefined
    );
  });

  it('should updateSearchResults again when submitting search field', async () => {
    const store = createMockedStore(defaultState);
    const app = mount(App, { store: new Vuex.Store(store) });

    store.modules.app.actions.updateSearchResults.mockClear();

    app.find(SearchField).vm.$emit('search');

    expect(store.modules.app.actions.updateSearchResults).toHaveBeenCalledWith(
      expect.anything(), undefined, undefined
    );
  });

  it('should showOneMorePage when scroll end', async () => {
    const store = createMockedStore(defaultState);
    const app = mount(App, { store: new Vuex.Store(store) });

    app.find(ScrollTrackingContainer).vm.$emit('scroll-end');

    expect(store.modules.app.actions.showOneMorePage).toHaveBeenCalledWith(
      expect.anything(), undefined, undefined
    );
  });
});
