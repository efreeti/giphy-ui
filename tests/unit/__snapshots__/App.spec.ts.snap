// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div class="md-app md-app-side-drawer md-layout-row md-fixed md-theme-default">
  <main class="md-app-container md-flex md-layout-column md-theme-default md-scrollbar" style="padding-left: 0px;">
    <div class="md-toolbar md-app-toolbar md-layout md-alignment-center-center md-theme-default md-elevation-4" style="top: 0px;">
      <div class="md-layout-item md-size-40">
        <div class="md-field md-theme-default"><label for="md-input-zolfdnfyf">Search keyword(s)</label> <input type="text" id="md-input-zolfdnfyf" class="md-input"> <button type="button" class="md-button md-icon-button md-dense md-theme-default">
            <div class="md-ripple">
              <div class="md-button-content"><i class="md-icon md-icon-font md-theme-default">search</i></div>
            </div>
          </button>
          <!---->
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <div class="md-app-scroller md-layout-column md-flex md-theme-default md-scrollbar">
      <div class="md-content md-app-content md-flex md-layout md-alignment-center-center md-theme-default">
        <div class="md-layout-item md-size-40">
          <div>
            <div>
              <div>
                <div class="image-card"><span class="image-background"><span><div role="placeholder" style="width: 100px; height: 100px;"></div></span></span></div>
              </div>
              <div>
                <div class="image-card"><span class="image-background"><span><div role="placeholder" style="width: 100px; height: 100px;"></div></span></span></div>
              </div>
              <div>
                <div class="image-card"><span class="image-background"><span><div role="placeholder" style="width: 100px; height: 100px;"></div></span></span></div>
              </div>
              <div>
                <div class="image-card"><span class="image-background"><span><div role="placeholder" style="width: 100px; height: 100px;"></div></span></span></div>
              </div>
              <div>
                <div class="image-card"><span class="image-background"><span><div role="placeholder" style="width: 100px; height: 100px;"></div></span></span></div>
              </div>
              <!---->
            </div>
            <div>
              <div>
                <div>
                  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
                </div></span>
              </div></span>
            </div>
          </div>
          <div>
            <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
          </div></span>
        </div></span>
      </div>
    </div>
    <div>
      <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
    </div></span>
</div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"><div><div><i class="md-icon md-icon-font md-size-5x md-accent md-theme-default">error</i></div> <div><span class="error-message-text md-body-2">
            Something went wrong, try refreshing the page
          </span></div>
</div></span></div></span></div>
</div>
</div>
</div>
<div>
  <div>
    <div>
      <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
    </div>
  </div>
  <div>
    <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
  </div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
<div>
  <div class="image-card"><span class="image-background"><div class="image-placeholder"><span class="image-placeholder-content"></span></div></span>
</div>
</div>
</div>
</div>
<div>
  <div role="scroll-end-tracker" class="scroll-end-tracker"></div>
</div>
</div>
</div>
</div>
</div>
</main>
</div>
`;
